api/utils/requestProcessor.js

```diff
@@ -0,0 +1,1159 @@
+            return common.returnMessage(params, 200, 'Request ignored: ' + params.cancelRequest);
+        }
+        common.log("request").i('Request ignored: ' + params.cancelRequest, params.req.url, params.req.body);
+    }
+};
+
+/**
+ * Process Request Data
+ */
+const processRequestData = (params, app) => {
+    plugins.dispatch("/i", {params: params, app: app});
+
+    if (params.qstring.events) {
+        if (params.promises)
+            params.promises.push(countlyApi.data.events.processEvents(params));
+        else
+            countlyApi.data.events.processEvents(params);
+    } else if (plugins.getConfig("api").safe && !params.bulk) {
+        return common.returnMessage(params, 200, 'Success');
+    }
+
+    if (countlyApi.data.usage.processLocationRequired(params)) {
+        countlyApi.data.usage.processLocation(params).then(() => continueProcessingRequestData(params, () => {}));
+    } else {
+        continueProcessingRequestData(params, () => {});
+    }
+};
+
+/**
+ * Continue Processing Request Data
+ * @returns {boolean}
+ */
+const continueProcessingRequestData = (params, done) => {
+    if (params.qstring.begin_session) {
+        countlyApi.data.usage.beginUserSession(params, done);
+    } else {
+        if (params.qstring.metrics) {
+            countlyApi.data.usage.processMetrics(params);
+        }
+        if (params.qstring.end_session) {
+            if (params.qstring.session_duration) {
+                countlyApi.data.usage.processSessionDuration(params, () => {
+                    countlyApi.data.usage.endUserSession(params, done);
+                });
+            } else {
+                countlyApi.data.usage.endUserSession(params, done);
+            }
+        } else if (params.qstring.session_duration) {
+            countlyApi.data.usage.processSessionDuration(params, () => {
+                return done ? done() : false;
+            });
+        } else {
+            //update lac, all other requests do updates to app_users and update lac automatically
+            common.updateAppUser(params, {$set: {lac: params.time.mstimestamp}});
+
+            // begin_session, session_duration and end_session handle incrementing request count in usage.js
+            const dbDateIds = common.getDateIds(params),
+                updateUsers = {};
+
+            common.fillTimeObjectMonth(params, updateUsers, common.dbMap['events']);
+            const postfix = common.crypto.createHash("md5").update(params.qstring.device_id).digest('base64')[0];
+            common.db.collection('users').update({
+                '_id': params.app_id + "_" + dbDateIds.month + "_" + postfix
+            }, {'$inc': updateUsers}, {'upsert': true}, (err, res) => {
+            });
+
+            return done ? done() : false;
+        }
+    }
+};
+
+/**
+ * Process Bulk Request
+ * @param i
+ * @param requests
+ * @param params
+ * @param req
+ */
+const processBulkRequest = (i, requests, params, req) => {
```

Comment:
I think it actually only need i and params and the rest can be taken from params